test/e2e/gatewayapi_upgradeable_test.go

```diff
@@ -0,0 +1,89 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	gatewayapi_upgradeable "github.com/openshift/cluster-ingress-operator/pkg/operator/controller/gatewayapi-upgradeable"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	"github.com/openshift/api/features"
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+func kubeinit() {
+	kubeConfig, err := config.GetConfig()
+	if err != nil {
+		panic(err)
+	}
+	_, err = kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func TestGatewayAPIUpgradeable(t *testing.T) {
+	if gatewayAPIEnabled, err := isFeatureGateEnabled(features.FeatureGateGatewayAPI); err != nil {
+		t.Fatalf("error checking feature gate enabled status: %v", err)
+	} else if gatewayAPIEnabled {
+		t.Skip("Gateway API is enabled, skipping TestGatewayAPIUpgradeable")
+	}
+
+	t.Parallel()
+
+	createCRDs(t)
+	testAdminGate(t, true)
+	deleteCRDs(t)
+	testAdminGate(t, false)
+}
+
+func createCRDs(t *testing.T) {
+	crds := []*apiextensionsv1.CustomResourceDefinition{
+		manifests.GatewayClassCRD(),
+		manifests.GatewayCRD(),
+		manifests.HTTPRouteCRD(),
+		manifests.ReferenceGrantCRD(),
+	}
+
+	for _, crd := range crds {
+		if err := kclient.Create(context.TODO(), crd); err != nil {
+			if !errors.IsAlreadyExists(err) {
+				t.Fatalf("Failed to create CRD %s: %v", crd.Name, err)
+			}
+			continue
+		}
+	}
+}
+
+func testAdminGate(t *testing.T, shouldExist bool) {
+	t.Helper()
+	//	crdsExist := checkCRDsExist(t)
+	configMap := &corev1.ConfigMap{}
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 30*time.Second, false, func(ctx context.Context) (bool, error) {
+		if err := kclient.Get(ctx, client.ObjectKey{Namespace: "openshift-config-managed", Name: "admin-gates"}, configMap); err != nil {
+			t.Logf("Failed to get configmap admin-gates: %v, retrying...", err)
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Errorf("Timed out trying to get admin-gates configmap: %v", err)
+		return
+	}
+
+	_, adminGateKeyExists := configMap.Data[gatewayapi_upgradeable.GatewayAPIAdminKey]
```

Comment:
The tests should avoid using the same variables as used in the code they test. Otherwise we risk to repeat mistakes made in the code.
```suggestion
	_, adminGateKeyExists := configMap.Data["ack-gateway-api-management"]
```

And please unexport the admin key and admin message variables from the controller package.

Reply:
Fixed